### PR TITLE
Put `jl_gc_new_weakref` in a header file again

### PR DESCRIPTION
**Pull request by lgoettgens**
The gcext implementation in [gap](https://github.com/gap-system/gap/)/[GAP.jl](https://github.com/oscar-system/GAP.jl) uses `jl_gc_new_weakref`. Since https://github.com/JuliaLang/julia/pull/55608 got merged, this symbol is no longer mentioned in any header file, and thus the compilation of gap fails. This PR puts it back into a header file, but now in `gc-common.h` (instead of `gc-interface.h`), since https://github.com/JuliaLang/julia/pull/55608 moved the implementation to `gc-common.c`.

We are aware of the alternative function `jl_gc_new_weakref_th` (that still is in a header after https://github.com/JuliaLang/julia/pull/55608). However, it was not present in any header file before https://github.com/JuliaLang/julia/pull/55256, even though it was already marked as `JL_DLLEXPORT` well before that.

In short, without this PR we are not able to adapt the code to a way that works both with julia 1.11 and nightly.

@fingolfin I get locally still the same error as we can observe in CI. I suppose that this needs a rebuild of `libjulia` with this PR here included to work, right? 

ping @d-netto @gbaraldi

### CODE REVIEW

**Review by gbaraldi — APPROVED** [review]

**Comment by fingolfin** [comment]
Thank you!

Yes, libjulia needs the updated header. Since I only rebuilt it a few days ago, it should be trivial enough to do it again once this PR is merged

**Review by fingolfin — APPROVED** [review]

**Review by d-netto — APPROVED** [review]
LGTM. Thanks for spotting it.

File: src/gc-common.h [hunk 1]
```diff
@@ -185,4 +185,13 @@ extern jl_ptls_t* gc_all_tls_states;
...
 extern int gc_logging_enabled;
 
+// =========================================================================== //
+// Misc
+// =========================================================================== //
+
+// Allocates a new weak-reference, assigns its value and increments Julia allocation
+// counters. If thread-local allocators are used, then this function should allocate in the
+// thread-local allocator of the current thread.
+JL_DLLEXPORT struct _jl_weakref_t *jl_gc_new_weakref(struct _jl_value_t *value);
```
Comment (d-netto) [inline src/gc-common.h:195]: Minor style suggestion: since this is in `gc-common.h`, we should have access to `jl_weakref_t` and `jl_value_t`.

Perhaps use those, instead of `struct _jl_weakref_t` and `struct _jl_value_t`?

Comment (lgoettgens) [reply src/gc-common.h:195]:
```suggestion
JL_DLLEXPORT jl_weakref_t *jl_gc_new_weakref(jl_value_t *value);
```
Yeah, should work. @fingolfin do you agree?

**Comment by fingolfin** [comment]
Unfortunately this is not quite enough, as just `#include "julia.h"` still won't make this function visible again: Apparently `julia.h` only includes `src/gc-interface.h` but not `src/gc-common.h`.
